### PR TITLE
WIP: Try -Wunreachable code

### PR DESCRIPTION
**Pull request by richsalz**
I don't know that we want to do this.  Seeking opinions.
This doesn't solve *all* unreachable-code issues, but the majority of them.
Maybe some parts -- ALWAYS, SET_UPLINK for example -- are worth doing?

### CODE REVIEW

**Comment by dot-asm** [comment]
> I don't know that we want to do this.

What a trick question :-) We do have constant conditions, i.e. the `if(const) {} else {}` kind of thing, but those are totally intentional. `else`-clauses are customarily legacy code and common rationale is to expose it to analysis performed by compiler. I mean even if binary code is omitted, source code is still at least parsed to ensure that it doesn't rot in pre-processor `#else`. Bottom line is that we don't have it as goal to get rid of unreachable code. Well, of course, the suggestion is actually not to remove such code, but to *mask* warning in cases when unreachable code is intentional. What is unfortunate is that masking mechanism is a compiler-specific *quirk*, more specifically making a constant to an expression by surrounding it with braces. And it being a quirk kind of puts uncertainty in the value of the suggested effort. I mean question is if we want to rely on a quirk. Of course one can argue that using a quirk is better than nothing, and if it disappears at some future point, we can deal with it *then*... In other words I would *not* say that it's meaningless, yet somewhat dubious...

On side note one should also note that that particular clang warning might need more work, for example it apparently doesn't seem to recognize `if(sizeof(int)==4)` as constant condition, i.e. doesn't warn you that `else` to such condition is unreachable...

As for styling. I personally would find `if ((1))` more readable that `if (ALWAYS)`. Because latter still leaves question about what is "ALWAYS" and where is it defined/declared, while former leaves *no* questions whatsoever...

**Comment by FdaSilvaYY** [comment]
Hi, Rich.
I do some tryout in a branch sometimes ago : 
https://github.com/FdaSilvaYY/openssl/tree/clang-unreachable-code
I noticed two differences : 
- I didn't fix b_addr.c in the same way.
- You didn't fix a few tests on defines like `if (P256_LIMBS == 8)` in ecp_nistz256.c

**Comment by dot-asm** [comment]
There is a number of constant ifs that are totally intentional, *including* b_addr,c and ecp_nistz256.c. Speaking of warnings in ecp_nistz. Note that `if(P256_LIMBS==8)` doesn't do *despite* the fact that P256_LIMBS is defined as `(something)`, braces has to be around 8... As already said, it's quirky...

**Comment by FdaSilvaYY** [comment]
@dot-asm : I don't get what you mean by 
> ... doesn't do despite ...

For my point of view, this code ( b_addr,c and ecp_nistz256.c ) contains some values/conditions defined and already-known at compile-time .
Why are they expressed/deferred as execution-time tests ? I don't know.

**Comment by dot-asm** [comment]
> I don't get what you mean by "... doesn't do despite ..."

I meant that I would expect `if (P256_LIMBS==8)` to be warning-free because `P256_LIMBS` is an "expression" in round braces. But no, it insists on `if (P256_LIMBS==(8))`... I mean why not say triple-braces all over then? :-)

**Comment by dot-asm** [comment]
> code ( b_addr,c and ecp_nistz256.c ) contains some values/conditions defined and already-known at compile-time. Why are they expressed/deferred as execution-time tests ?

Not execution, but compile time. Optimizing compiler will routinely *remove* `if(0){code}`, so that it doesn't really matter if it's there or not. But is there any upside? Yes. For most of us platform of choice is same, hence there is tendency for blind spots. Or rather there would be blind spots if for example those pieces of code were in #ifdef. But if it's in `if(0){code}` instead, then at least typos and warnings are taken care of *even* on your platform of choice...

File: crypto/bio/bss_file.c [hunk 1]
```diff
@@ -92,8 +92,8 @@ BIO *BIO_new_file(const char *filename, const char *mode)
         return (NULL);
     }
 
-    BIO_clear_flags(ret, BIO_FLAGS_UPLINK); /* we did fopen -> we disengage
-                                             * UPLINK */
+    /* we did fopen -> we disengage UPLINK */
+    BIO_clear_flags(ret, BIO_FLAGS_UPLINK);
```
Comment (kaduk) [inline crypto/bio/bss_file.c]: This looks like just style; should it be a separate commit?

File: crypto/bio/bss_file.c [hunk 2]
```diff
@@ -300,8 +299,8 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
         }
         b->ptr = fp;
         b->init = 1;
-        BIO_clear_flags(b, BIO_FLAGS_UPLINK); /* we did fopen -> we disengage
-                                               * UPLINK */
+        /* we did fopen -> we disengage UPLINK */
+        BIO_clear_flags(b, BIO_FLAGS_UPLINK);
```
Comment (kaduk) [inline crypto/bio/bss_file.c]: Same here about style-only

File: crypto/include/internal/cryptlib.h [hunk 3]
```diff
@@ -17,7 +17,9 @@
 
 # ifdef OPENSSL_USE_APPLINK
 #  undef BIO_FLAGS_UPLINK
+#  undef BIO_UPLINK_SET
 #  define BIO_FLAGS_UPLINK 0x8000
+#  define BIO_UPLINK_SET(b)     (((b) & BIO_FLAGS_UPLINK) != 0)
```
Comment (kaduk) [inline crypto/include/internal/cryptlib.h]: Shouldn't this be (b)->flags?

File: crypto/bio/bss_file.c (hunk 1 above)
Comment (richsalz) [inline crypto/bio/bss_file.c]: Since I squash/merge, not worth the trouble.

File: crypto/include/internal/cryptlib.h (hunk 3 above)
Comment (richsalz) [inline crypto/include/internal/cryptlib.h]: yes.  oops.

**Comment by richsalz** [comment]
this is even *less* than WIP, it's more RFC. :)

**Comment by richsalz** [comment]
not worth doing.
